editoast/src/views/timetable/similar_trains/trains_traffic.rs

```diff
@@ -0,0 +1,361 @@
+    pub fn add_train_traffic(&mut self, train_data: TrainTraffic) -> Result<()> {
+        // Only import train that have a valid start date
+        if train_data.start_time > self.traffic_valid_from {
+            // Index the train by its speedlimit
+            self.trains_by_speedlimit
+                .entry(train_data.speed_limit_tag.clone())
+                .or_default()
+                .insert(train_data.id);
+
+            // Index the train by its rollingstock
+            self.trains_by_rollingstock
+                .entry(train_data.rolling_stock.clone())
+                .or_default()
+                .insert(train_data.id);
+
+            // Index the train by segment
+            for (from, to) in train_data
+                .operational_points
+                .iter()
+                .filter(|op| op.stop)
+                .tuple_windows()
+            {
+                self.trains_by_segment
+                    .entry(Segment {
+                        from: from.clone(),
+                        to: to.clone(),
+                    })
+                    .or_default()
+                    .insert(train_data.id);
+            }
+
+            // Index the train by its id
+            if self.trains_by_id.contains_key(&train_data.id) {
+                tracing::warn!(
+                    train_id = train_data.id,
+                    "Train already exist, overriding it"
+                );
+            }
+            self.trains_by_id.insert(train_data.id, train_data);
+        }
+
+        Ok(())
+    }
+
+    /// Find in the traffic a composition of trains that match the given train itinerary
+    pub fn find_compatible_trains(
+        &self,
+        rolling_stock: Option<String>,
+        speed_limit_tag: Option<String>,
+        waypoints: Vec<Waypoint>,
+    ) -> Vec<TrainTraffic> {
+        let mut valid_trains = HashSet::<usize>::new();
+
+        // Getting a list of valid train Ids which match the train's specifications
+        let compliant_trains = match (rolling_stock, speed_limit_tag) {
+            (Some(rs), None) => self
+                .trains_by_rollingstock
+                .get(&rs)
+                .cloned()
+                .unwrap_or_default(),
+            (None, Some(speed)) => self
+                .trains_by_speedlimit
+                .get(&speed)
+                .cloned()
+                .unwrap_or_default(),
+            (Some(rs), Some(speed)) => {
+                let rs_trains = self
+                    .trains_by_rollingstock
+                    .get(&rs)
+                    .cloned()
+                    .unwrap_or_default();
+                let speed_trains = self
+                    .trains_by_speedlimit
+                    .get(&speed)
+                    .cloned()
+                    .unwrap_or_default();
+                rs_trains.intersection(&speed_trains).cloned().collect()
+            }
+            (None, None) => self.trains_by_id.keys().cloned().collect(),
```

Comment:
The API does not support searching the whole database without specifying the rolling stock or the speed limit tag. This should be unreachable!.

Reply:
Yes but in this function's  argument, it's possible to have both RS & Speed tag empty, so we need to handle it in one way or an other.  
Adding an another check and raise an error here is redundant, I prefer let upper code handles the business code.

Reply:
I meant using the `unreachable!` macro which is meant for this use case.

```suggestion
            (None, None) => unreachable!("should be checked at API-level"),
```

Reply:
I'd suggest 
```rust
// Getting a list of valid train Ids which match the train's specifications
let empty = HashSet::<usize>::new();
let compliant_trains = match (rolling_stock, speed_limit_tag) {
    (Some(rs), None) => self.trains_by_rollingstock.get(&rs).unwrap_or(&empty),
    (None, Some(speed)) => self.trains_by_speedlimit.get(&speed).unwrap_or(&empty),
    (Some(rs), Some(speed)) => {
        let rs_trains = self.trains_by_rollingstock.get(&rs).unwrap_or(&empty);
        let speed_trains = self.trains_by_speedlimit.get(&speed).unwrap_or(&empty);
        &rs_trains.intersection(speed_trains).copied().collect()
    }
    (None, None) => unreachable!("Should not be reachable, checked at the API level"),
};
```

to make it a bit less clony, since it's my understanding that the HashSets are potentially large.